File: 1923-sentence-similarity-iii/sentence-similarity-iii.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

// Helper function to split sentence into words
char** splitSentence(char* sentence, int* wordCount) {
    *wordCount = 1;
    for (int i = 0; sentence[i]; i++) {
        if (sentence[i] == ' ') (*wordCount)++;
    }
    
    char** words = (char**)malloc(*wordCount * sizeof(char*));
    char* token = strtok(sentence, " ");
    int index = 0;
    
    while (token != NULL) {
        words[index] = strdup(token);
        token = strtok(NULL, " ");
        index++;
    }
    
    return words;
}

bool areSentencesSimilar(char* sentence1, char* sentence2) {
    // Create copies of sentences as strtok modifies the original string
    char* s1_copy = strdup(sentence1);
    char* s2_copy = strdup(sentence2);
    
    int count1, count2;
    char** words1 = splitSentence(s1_copy, &count1);
    char** words2 = splitSentence(s2_copy, &count2);
    
    // Ensure sentence1 is the shorter one
    if (count1 > count2) {
        char** tempWords = words1;
        words1 = words2;
        words2 = tempWords;
        int tempCount = count1;
        count1 = count2;
        count2 = tempCount;
    }
    
    int left = 0;
    // Match words from left
    while (left < count1 && strcmp(words1[left], words2[left]) == 0) {
        left++;
    }
    
    int right = 0;
    // Match words from right
    while (right < count1 - left && 
           strcmp(words1[count1 - 1 - right], 
                 words2[count2 - 1 - right]) == 0) {
        right++;
    }
    
    // Free allocated memory
    for (int i = 0; i < count1; i++) free(words1[i]);
    for (int i = 0; i < count2; i++) free(words2[i]);
    free(words1);
    free(words2);
    free(s1_copy);
    free(s2_copy);
    
    return (left + right) >= count1;
}

File: 1923-sentence-similarity-iii/sentence-similarity-iii.java
class Solution {
    public boolean areSentencesSimilar(String sentence1, String sentence2) {
        // Split sentences into words
        String[] words1 = sentence1.split(" ");
        String[] words2 = sentence2.split(" ");
        
        // Ensure words1 is the shorter array
        if (words1.length > words2.length) {
            String[] temp = words1;
            words1 = words2;
            words2 = temp;
        }
        
        int n1 = words1.length;
        int n2 = words2.length;
        
        int left = 0;
        // Match words from left
        while (left < n1 && words1[left].equals(words2[left])) {
            left++;
        }
        
        int right = 0;
        // Match words from right
        while (right < n1 - left && 
               words1[n1 - 1 - right].equals(words2[n2 - 1 - right])) {
            right++;
        }
        
        return (left + right) >= n1;
    }
    
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Test cases
        System.out.println("Test 1: " + 
            solution.areSentencesSimilar(
                "My name is Haley", 
                "My Haley")); // Should print true
        
        System.out.println("Test 2: " + 
            solution.areSentencesSimilar(
                "of", 
                "A lot of words")); // Should print true
        
        System.out.println("Test 3: " + 
            solution.areSentencesSimilar(
                "Eating right now", 
                "Eating")); // Should print true
                
        System.out.println("Test 4: " + 
            solution.areSentencesSimilar(
                "A B C", 
                "A D B C")); // Should print true
    }
}